software/firmware/europi.py

```diff
@@ -472,21 +474,57 @@ def __init__(
                 )
         super().__init__(self.width, self.height, i2c)
 
-    def centre_text(self, text):
-        """Split the provided text across 3 lines of display."""
+    def writer(self, font):
+        """Returns the large font writer for the specified font."""
+        n = font.__name__
+        if n not in self.writers:
+            self.writers[n] = Writer(self, font)
+        return self.writers[n]
+
+    def text_len(self, s, font=None):
+        """Returns the length of the string s in pixels."""
+        if font is None:
+            return len(s) * 8
+        else:
+            return self.writer(font).string_len(s)
+
+    def text(self, s, x, y, c=1, font=None):
+        """Display the string s using the x, y coordinates as the upper-left corner of the text.
+        When using a custom font, the text is not displayed if it does not fit in the display.
+        """
+        if font is None:
+            return super().text(s, x, y, c)
+        self.writer(font).print(s, x, y, c)
+
+    def centre_text(self, text, font=None):

```

Comment:
[required] As discussed in my main comment, let's move the changes to support experimental features out of `europi.py` until they have been proven.

Reply:
Yes, I agree.

How would you like this support to be made available? As a module in the experimental package or a module in the contrib package?
And I think we must leave the Display class as it is now and create a new DisplayCustom (naming things is hard) subclass which inherits from Display.
And maybe provide a couple of custom fonts and simply give the user instructions about how to create its own custom fonts.

Reply:
I suggested the same thing in a previous comment, glad to see the same idea here!

Reply:
> How would you like this support to be made available?

I think that both your and @awonak's thoughts here make sense to me, and line up with the approach I outlined in my initial comment [here](https://github.com/Allen-Synthesis/EuroPi/pull/234#pullrequestreview-1326475657).

As for a class name, maybe `CustomFontDisplay`?